### ISSUE TITLE
Build: Investigate what is required to support Springboot better

### ISSUE THREAD

**Issue by sthaha** [openshiftio/openshift.io#3942]

**Comment by lordofthejars** [comment]
This issue is really ambiguous, because what it means better? Now it works, so what I assume is what build team should do to support advanced cases.

Maybe one advanced case would be the communication between two components, `Component A -> Component B` where `Component B` can be another service or a database, but in any case, it is pretty similar. To simplify from now the scenario will be a Spring Boot service + JPA database. An example can be found at https://github.com/lordofthejars/jpa-onetomany-springboot-maven-mysql

Then steps to move (3000 ft view)

- Code

* Add Spring Data
* Add logic in app to use Spring Data
* Create a `db.sql` that creates the database schema
* Create a test with Arquillian APE

- Infrastructure 

* Create a directory `src/main/fabric8`
* Create `postgresql-deployment.yml` inside there where we define all postgresql required information such as pod and service, ...
* Add a custom liveness check to this directory for spring boot app so it says that it is ready when postgresql service is also ready.

Then when running `f-m-p` it will generate the resources for the spring boot app but also apply the resources placed at `src/main/fabric8` so the postgresql service will be started as well.

NOTE: I am not sure if we are not going to reach some kind of resource limit.

**Comment by bartoszmajsak** [comment]
I would start simple to see if you could handle service with a db and deploy that through pipeline. Probably first focus on the figuring out if such an app could be deployed on OSIO and only then start adding extras like APE tests.

If you encounter issues which are beyond "build" we should definitely reach out to other teams.

I really like @chmouel way of [evaluating NodeJS](https://hackmd.io/aOuvTaU2QM6raiwdTcFB0A) - maybe you could follow similar approach to capture all the pains on the way.

If you wish we could do some "pair programming" on that. Start looking at Spring Boot Boosters if there is something we can build upon.

**Comment by lordofthejars** [comment]
@bartoszmajsak Sure, let's do that it is a really good idea.

**Comment by hrishin** [comment]
> I would start simple to see if you could handle service with a DB and deploy that through the pipeline

From build standpoint that makes sense. It would be better to provision application dependencies (DB, cache, MQ, may be cloud resources) through pipeline, maybe using service catalog. (_real world app_).

**Comment by sthaha** [comment]
@lordofthejars 
> This issue is really ambiguous, because what it means better? Now it works, so what I assume is what build team should do to support advanced cases. 

Good point! I think @bmicklea who came up with this requirement will be able to help us define what a better support for SB will be like

**Comment by lordofthejars** [comment]
> It would be better to provision application dependencies (DB, cache, MQ, may be cloud resources) through pipeline

This is something that it is managed by kubernetes (case of definition + lifecycle), and then there is the "when it is started" which could be managed by pipeline as you suggested, but if fmp can (and I have used it) for this, then I think we should maintain pipeline simple. I will record a demo when I have everything in place.

**Comment by lordofthejars** [comment]
I have an initial version working of Spring Boot + DB in OSIO. I have taken an example of boosters and adapted to OSIO, then I  have contributed upstream.

Now working on the testing part.

**Comment by lordofthejars** [comment]
I have been working with this example and I have faced two problems that now need a deeper and broader analysis:

1) Quota problems: Now instead of 1 we are trying to release two pods every time. This means that you reach quota problems when you are trying to update the application. The first time you are releasing the application there is no problem since `staging` environment is empty, but when you update the application to a new version, you have not enough quota to deploy the new version. Even you get a dead lock situation where you need to update to a new version of the service but you have no quota to start it, but since you cannot scale up that service you cannot scale down the old version (to get a free slot) because of rolling strategy.

2) The second problem happens when you send a PR. In this case, the CI stage happens (instead of CD) so you try to run integration tests (aka `OpenshiftIT` tests). In this case we are trying to deploy the service into Openshift cluster, but you are trying to deploy to jenkins namespace. What's happening now is that since this namespace is already running all pods required by Jenkins, we have no permissions to deploy the service for testing purposes, so the test fails. `Service account may have been revoked. services "spring-boot-rest-http-crud" is forbidden: exceeded quota: object-counts, requested: services=1, used: services=5, limited: services=5.`

So as you can see there is a major problem regarding supporting more advanced use cases.

The third problem not related directly with this is that `OpenshiftIT` test would work only when it is executed from CLI (Maven) but not from Che (`Run As Test`) This problem is more related with the integration between FMP and Arquillian rather than Che, but of course we might have the same problem of quota depending on the number of pods required.

**Comment by pradeepto** [comment]
cc @gorkem @bmicklea @slemeur PTAL and comment.

**Comment by bartoszmajsak** [comment]
Just for traceability - there was some work needed in the booster itself to make it running https://github.com/snowdrop/spring-boot-crud-booster/pull/46 (ended up as a separated branch, hence this PR is closed, not mergeD)

**Comment by bmicklea** [comment]
For #1 could we stop using a rollover strategy? It's not really necessary
for staging at least is it?  Or is this only solvable by getting more RAM /
tenant?

**Comment by lordofthejars** [comment]
@bmicklea, changing the rolling strategy is just postponing the problem. We can change the rolling strategy then it will work in this concrete case, but when we move to another example (for example 2 services + Artemis) we will face the same problem again.

So I think that the solution might be by allowing the user to run more pods in namespaces so we can cover more use cases. Something similar would happen for example if we want to create VertX example communication with bus.

**Comment by bmicklea** [comment]
Makes sense but "allowing users to run more pods" means giving them more
RAM, right? Or does it require a different change?

**Comment by lordofthejars** [comment]
Not 100% sure, but I think it means more RAM and maybe CPU.
